crates/nu-protocol/src/errors/shell_error.rs

```diff
@@ -108,6 +107,34 @@ pub enum ShellError {
         span: Span,
     },
 
+    /// A value's type did not match the expected type.
+    ///
+    /// ## Resolution
+    ///
+    /// Convert the value to the correct type or provide a value of the correct type.
+    #[error("Type mismatch")]
+    #[diagnostic(code(nu::shell::type_mismatch))]
+    RuntimeTypeMismatch {
+        expected: Type,
+        actual: Type,
+        #[label = "expected {expected}, but got {actual}"]
+        span: Span,
+    },
```

Comment:
Mhh... #10698 

Definitely better than the existing trash-heap of `TypeMismatch`. Not sure if we should give this name provisionally to something with just one span.